SYCL/ESIMD/api/functional/ctors/ctor_default.cpp

```diff
@@ -0,0 +1,171 @@
+template <typename DataTypeT, int NumElems> struct test {
+  void operator()(const Logger &log, sycl::queue &queue, bool &passed) {
+    DataTypeT default_val{};
+    using AllocatorT = sycl::usm_allocator<DataTypeT, sycl::usm::alloc::shared>;
+
+    // verification for initializing simd by initializer
+    {
+      std::vector<DataTypeT, AllocatorT> result{NumElems, AllocatorT{queue}};
+      queue.submit([&](sycl::handler &cgh) {
+        auto out = result.data();
+        cgh.single_task<kernel_for_simd<DataTypeT, 1, NumElems>>([=] {
+          simd<DataTypeT, NumElems> simd_by_init = simd<DataTypeT, NumElems>();
+          simd_by_init.copy_to(out);
+        });
+      });
+      for (const auto &it : result) {
+        if (it != default_val) {
+          passed = false;
+          log.fail<DataTypeT, NumElems>(
+              "In simd by initializer elem value is not equal to default "
+              "value, "
+              "it value is: " +
+              std::to_string(it) +
+              ", but provided data type default value is: " +
+              std::to_string(default_val));
+        }
+      }
+    }
+
+    // verification for initializing simd by variable declarator
+    {
+      std::vector<DataTypeT, AllocatorT> result{NumElems, AllocatorT{queue}};
+      queue.submit([&](sycl::handler &cgh) {
+        auto out = result.data();
+        cgh.single_task<kernel_for_simd<DataTypeT, 2, NumElems>>([=] {
+          simd<DataTypeT, NumElems> simd_by_var_decl;
+          simd_by_var_decl.copy_to(out);
+        });
+      });
+      for (const auto it : result) {
+        if (it != default_val) {
+          passed = false;
+          log.fail<DataTypeT, NumElems>(
+              "In simd by variable declaration elem value is not equal to "
+              "default value, it value is: " +
+              std::to_string(it) +
+              ", but provided data type default value is: " +
+              std::to_string(default_val));
+        }
+      }
+    }
+
+    // verification for initializing simd by rvalue in an expression
+    {
+      std::vector<DataTypeT, AllocatorT> result{NumElems, AllocatorT{queue}};
+      queue.submit([&](sycl::handler &cgh) {
+        auto out = result.data();
+        cgh.single_task<kernel_for_simd<DataTypeT, 3, NumElems>>([=] {
+          simd<DataTypeT, NumElems> simd_by_rval;
+          simd_by_rval = simd<DataTypeT, NumElems>();
+          simd_by_rval.copy_to(out);
+        });
+      });
+      for (const auto &it : result) {
+        if (it != default_val) {
+          passed = false;
+          log.fail<DataTypeT, NumElems>(
+              "In simd by rvalue in an expression elem value is not equal to "
+              "default value, it value is: " +
+              std::to_string(it) +
+              ", but provided data type default value is: " +
+              std::to_string(default_val));
+        }
+      }
+    }
+
+    // verification for initializing simd by rvalue in actual parameter
```

Comment:
Looks like we have 4 blocks of code with minor differences here. Could we possibly have a templated version of this code, with some "action" and "description" passed?
So that:

- if we want to add 1 more "action" to test, we don't need to copy the entire logic
- if we want to update test logic, we don't need to update it in multiple places

Reply:
I've applied this comment in this commit: 35fd849b6fc222cb2da9d90a1dd6185de07651e7